platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/HttpRequest.java

```diff
@@ -0,0 +1,231 @@
+package com.mapbox.mapboxsdk.http;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.os.Build;
+import android.support.annotation.IntDef;
+import android.text.TextUtils;
+
+import com.mapbox.mapboxsdk.BuildConfig;
+import com.mapbox.mapboxsdk.Mapbox;
+import com.mapbox.mapboxsdk.constants.MapboxConstants;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.net.NoRouteToHostException;
+import java.net.ProtocolException;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.net.ssl.SSLException;
+
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.internal.Util;
+import timber.log.Timber;
+
+/**
+ * Gateway to request and return HttpRequest to c++
+ * <p>
+ * Reentrancy lock is not needed, but "Lock" is an abstract class.
+ * </p>
+ */
+class HttpRequest implements Callback {
+
+  @IntDef( {CONNECTION_ERROR, TEMPORARY_ERROR, PERMANENT_ERROR})
+  @Retention(RetentionPolicy.SOURCE)
+  @interface HttpError {
+  }
+
+  private static final int CONNECTION_ERROR = 0;
+  private static final int TEMPORARY_ERROR = 1;
+  private static final int PERMANENT_ERROR = 2;
+
+  private static final ReentrantLock LOCK = new ReentrantLock();
+  private static final OkHttpClient HTTP_CLIENT = new OkHttpClient();
+
+  private static boolean logEnabled = true;
+  private long nativePtr;
+  private Call call;
+
+  private native void nativeOnFailure(@HttpError int type, String message);
+
+  private native void nativeOnResponse(int code, String etag, String modified, String cacheControl, String expires,
+                                       String retryAfter, String xRateLimitReset, byte[] body);
+
+  private HttpRequest(long nativePtr, String resourceUrl, String etag, String modified) {
+    this.nativePtr = nativePtr;
+
+    try {
+      // Don't try a request if we aren't connected
+      if (!Mapbox.isConnected()) {
+        throw new NoRouteToHostException("No Internet connection available.");
+      }
+
+      HttpUrl httpUrl = HttpUrl.parse(resourceUrl);
+      resourceUrl = adaptResourceUrl(httpUrl, resourceUrl);
+
+      call = HTTP_CLIENT.newCall(buildRequest(resourceUrl, etag, modified));
+      call.enqueue(this);
+    } catch (Exception exception) {
+      onFailure(exception);
+    }
+  }
+
+  private String adaptResourceUrl(HttpUrl httpUrl, String resourceUrl) {
+    final String host = httpUrl.host().toLowerCase(MapboxConstants.MAPBOX_LOCALE);
+    if (host.equals("mapbox.com") || host.endsWith(".mapbox.com") || host.equals("mapbox.cn")
+      || host.endsWith(".mapbox.cn")) {
+      if (httpUrl.querySize() == 0) {
+        resourceUrl = resourceUrl + "?";
```

Comment:
This will be way cleaner:
```java

  private String adaptResourceUrl(String resourceUrl) {
    HttpUrl httpUrl = HttpUrl.parse(resourceUrl);
    if (httpUrl != null && isMapboxHost(httpUrl)) {
      HttpUrl.Builder builder = httpUrl.newBuilder();
      builder.addQueryParameter("events", "true");
      resourceUrl = builder.build().toString();
    }
    return resourceUrl;
  }
```